FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/SchedulerService.cs

```diff
@@ -334,91 +428,142 @@ private async Task CreateNextTriggerAsync(CurrentTriggerEntity currentTriggerEnt
+                    bool isSucceeded = await TryRenewLeaseInternalAsync(cancellationToken);
+                    if (!isSucceeded)
                     {
-                        _logger.LogInformation("Failed to renew lease, the retrieved lease trigger entity is null or working instance doesn't match.");
+                        delayTaskCancellationTokenSource.Cancel();
+                        await intervalDelayTask;
+                        break;
                     }
-                    else
-                    {
-                        triggerLeaseEntity.HeartbeatDateTime = DateTimeOffset.UtcNow;
 
-                        await _metadataStore.UpdateEntityAsync(
-                            triggerLeaseEntity,
-                            cancellationToken: cancellationToken);
-                    }
+                    await intervalDelayTask;
                 }
                 catch (Exception ex)
                 {
-                    _logger.LogInformation(ex, $"Failed to renew lease for working instance {_instanceGuid}.");
-                }
-
-                if (!cancellationToken.IsCancellationRequested)
-                {
+                    _logger.LogInformation(ex, $"Scheduler instance {_instanceGuid} fails to renew lease.");
+                    delayTaskCancellationTokenSource.Cancel();
                     await intervalDelayTask;
+                    break;
                 }
             }
 
-            _logger.LogInformation($"Stop to renew lease for working instance {_instanceGuid}.");
+            _logger.LogInformation($"Scheduler instance {_instanceGuid} stops to renew lease in loop.");
+        }
+
+        /// <summary>
+        /// try to renew lease
+        /// </summary>
+        /// <param name="cancellationToken">cancellation token.</param>
+        /// <returns>return true if the lease is renewed, otherwise return false or throw unknown exceptions</returns>
+        private async Task<bool> TryRenewLeaseInternalAsync(CancellationToken cancellationToken)
+        {
+            // try to get trigger lease entity
+            TriggerLeaseEntity triggerLeaseEntity =
+                await _metadataStore.GetTriggerLeaseEntityAsync(_queueType, cancellationToken);
+
+            if (triggerLeaseEntity == null)
+            {
+                _logger.LogInformation($"Scheduler instance {_instanceGuid} fails to renew lease, the retrieved lease trigger entity is null.");
+                return false;
+            }
+
+            if (triggerLeaseEntity.WorkingInstanceGuid != _instanceGuid)
+            {
+                _logger.LogInformation($"Scheduler instance {_instanceGuid} Fails to renew lease, the retrieved lease trigger entity's working instance {triggerLeaseEntity.WorkingInstanceGuid} doesn't match.");
+                return false;
+            }
+
+            triggerLeaseEntity.HeartbeatDateTime = DateTimeOffset.UtcNow;
+
+            await _metadataStore.TryUpdateEntityAsync(
```

Comment:
Is it possible return false here?

Reply:
Yes, it is possible return false, fixed.